EIPS/eip-2771.md

```diff
@@ -0,0 +1,185 @@
+2. Extract the **Transaction Signer** address from the last 20 bytes of the
+   call data and use that as the original `sender` of the transaction (instead of `msg.sender`)
+3. If the `msg.sender` is not a trusted forwarder (or if the msg.data is
+   shorter than 20 bytes), then return the original `msg.sender` as it is.
+
+The **Recipient** MUST check that it trusts the Forwarder to prevent it from
+extracting address data appended from an untrusted contract. This could result
+in a forged address.
+
+### Protocol Support Discovery Mechanism
+
+Unless a **Recipient** contract is being used by a particular frontend that
+knows that this contract has support for native meta transactions, it would not
+be possible to offer the user the choice of using meta-transaction to interact
+with the contract. We thus need a mechanism by which the **Recipient** can let
+the world know that it supports meta transactions. 
+
+This is especially important for meta transactions to be supported at the Web3
+wallet level. Such wallets may not necessarily know anything about the
+**Recipient** contract users may wish to interact with.
+
+As a **Recipient** could trust forwarders with different interfaces and
+capabilities (e.g., transaction batching, different message signing formats),
+we need to allow wallets to discover which Forwarder is trusted.
+
+To provide this discovery mechanism a **Recipient** contract MUST implement
+this function:
+
+```solidity
+function isTrustedForwarder(address forwarder) external returns(bool);
+```
+
+*  That function MUST return true if the forwarder is trusted by the
+   Recipient.
+*  That function MUST return false if the forwarder is not trusted.
+*  That function MUST NOT throw a revert.
+
+Internally, the **Recipient** MUST then accept a request from forwarder
+
+That function can be called on-chain and as such gas restriction needs to be
+put in place. 
+
+A Gas limit of 10k is enough for making  the decision either inside the
+contract, or delegating it to another contract and doing some memory access
+calculations, like querying a mapping.
+
+###  Recipient example 
+
+```solidity
+contract RecipientExample {
+
+    function purchaseItem(uint256 itemId) external {
+        address sender = _msgSender();
+        ... perform the purchase for sender
+    }
+
+    address immutable _trustedForwarder;
+    constructor(address trustedForwarder) internal {
+        _trustedForwarder = trustedForwarder;
+    }
+
+    function isTrustedForwarder(address forwarder) external returns(bool) {
+        return forwarder == _trustedForwarder;
+    }
+
+    function _msgSender() internal view returns (address payable signer) {
+        signer = msg.sender;
+        if (isTrustedForwarder(signer)) {
+            bytes memory data = msg.data;
+            uint256 length = msg.data.length;
+            assembly { signer := mload(add(data, length))) }
+        }    
+    }
+
+}
+```
+
+## Security Considerations
```

Comment:
Missing Rationale section.  This is where you indicate why you made specific decisions you did in the specification.  Like why choose name X over name Y, why fire events at point A instead of point B, etc.
```suggestion
## Rationale
TBD

## Security Considerations
```

Reply:
Added Rationale section.